### PR TITLE
Merge ansys-dpf-core 0.4.0 from private repository to GitHub

### PR DESCRIPTION
**Pull request by cbellot000**
Make ansys-dpf-core compatible with Ansys release 2022R1
Update master version to 0.4.dev0

> Enhancements:
> - native remote workflows capability: [Workflows](https://dpfdocs.pyansys.com/api/ansys.dpf.core.workflow.html) and [Operators](https://dpfdocs.pyansys.com/api/ansys.dpf.core.dpf_operator.html) instantiated on different servers can be connected together transparently to create distributed pre and post processing workflows. Workflows created on a given server can be moved to other servers.
> - add optional progress bar when [Operators](https://dpfdocs.pyansys.com/api/ansys.dpf.core.dpf_operator.html) or [Workflows](https://dpfdocs.pyansys.com/api/ansys.dpf.core.workflow.html) are evaluated. This uses the new Session API through grpc
> - server on Docker Container is started automatically with, either the environment variable DPF_DOCKER or either the parameter `docker_name` of the function [`start_local_server`](https://dpfdocs.pyansys.com/api/ansys.dpf.core.server.html) set to the name of the docker. Unit tests can be run on Docker Container
> - implement Integral Collection to connect and get output data as int, double or string arrays more efficiently from [Operators](https://dpfdocs.pyansys.com/api/ansys.dpf.core.dpf_operator.html) and [Workflows](https://dpfdocs.pyansys.com/api/ansys.dpf.core.workflow.html)
> - create generic calls to get and set field properties from and to the [MeshedRegion](https://dpfdocs.pyansys.com/api/ansys.dpf.core.meshed_region.html)
> - identify DPF entities in grpc databases by their entity IDs and their DataBase IDs to allow for better communication between servers
> - add capability to generate [operators](https://dpfdocs.pyansys.com/api/ansys.dpf.core.operators.html) python code client side
> - by default, the mesh is read when a displacement is requested (can be disabled with [`model.mesh_by_default = False`](https://dpfdocs.pyansys.com/api/ansys.dpf.core.model.html) )
> - new operators to support [averaging](https://dpfdocs.pyansys.com/api/ansys.dpf.core.operators.averaging.html) from elemental centroids results
> - improve handling of MAPDL coupled analysis 
> - create [path_utilities](https://dpfdocs.pyansys.com/api/ansys.dpf.core.operators.path_utilities.html) to offer helpers on path depending on the server os

> Bug fixes:
> - automatic code generation for [`model.results....`](https://dpfdocs.pyansys.com/api/ansys.dpf.core.results.html) is working for results with scripting names having " ", "/", "\", ".", ":", ";", ",", "(", ")" characters.
> - decorator `server_meet_version` "wraps" the method so that the call stack when exception are raised is correct
> - fetch named selections from the [MeshedRegion](https://dpfdocs.pyansys.com/api/ansys.dpf.core.meshed_region.html) only when it's specifically required to improve performances
> - properly release memory server side when time/frequencies are requested by the [TimeFreqSupport](https://dpfdocs.pyansys.com/api/ansys.dpf.core.time_freq_support.html)
> - add more unary calls in [ResultInfo](https://dpfdocs.pyansys.com/api/ansys.dpf.core.result_info.html) service

> Documentation fixes:
> - README.md reference pydpf-core instead of DPF-Core
> - improve spelling

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -26,6 +26,12 @@ Python environment.
 
 ## Documentation
 
+Visit the [DPF-Core Documentation](https://dpfdocs.pyansys.com) for a
+detailed description of the library, or see the [Examples
+Gallery](https://dpfdocs.pyansys.com/examples/index.html) for more
+detailed examples.
+
+
```
Comment (akaszynski) [inline README.md:34]: Duplicate
```suggestion
```

File: ansys/dpf/core/__init__.py [hunk 2]
```diff
@@ -62,35 +62,28 @@
 from ansys.dpf.core import server
 from ansys.dpf.core import check_version
 from ansys.dpf.core import settings
+from ansys.dpf.core import path_utilities
+from ansys.dpf.core import settings
```
Comment (akaszynski) [inline ansys/dpf/core/__init__.py:67]: These really should be combined in parenthesis and merged with above.

File: ansys/dpf/core/_version.py [hunk 3]
```diff
@@ -1,8 +1,8 @@
 """Version for ansys-dpf-core"""
 # major, minor, patch
-version_info = 0, 3, "dev0"
+version_info = 0, 4, '0'
```
Comment (akaszynski) [inline ansys/dpf/core/_version.py:3]: Version tag should be `"dev0"` on `master` branch. Non dev versions should be on the release branches.

File: ansys/dpf/core/collection.py [hunk 4]
```diff
@@ -59,9 +61,43 @@ def __init__(self, dpf_type, collection=None, server: server.DpfServer = None):
...
+
         else:
             self._message = collection
 
+        if self._type == None:
+            self._type = types(int(self._message.type) + 1)
+
+    @staticmethod
+    def integral_collection(inpt, server: server.DpfServer = None):
+        """Creates a collection of integral type with a list.
```
Comment (akaszynski) [inline ansys/dpf/core/collection.py:73]: Short summary and long description should have a single line between.
```suggestion
        """Creates a collection of integral type with a list.

```

File: ansys/dpf/core/collection.py [hunk 5]
```diff
@@ -59,9 +61,43 @@ def __init__(self, dpf_type, collection=None, server: server.DpfServer = None):
...
+        The collection of integral is the equivalent of an array of
+        data sent server side. It can be used to efficiently stream
+        large data to the server.
+
+        Notes
+        -----
+        Used by default by the ``'Operator'`` and the``'Workflow'`` when a
+        list is connected or returned.
```
Comment (akaszynski) [inline ansys/dpf/core/collection.py:81]: Notes should be after the Returns section. See https://numpydoc.readthedocs.io/en/latest/format.html

File: ansys/dpf/core/collection.py [hunk 6]
```diff
@@ -162,6 +199,26 @@ def _get_entries(self, label_space_or_index):
         entries : list[Scoping], list[Field], list[MeshedRegion]
             Entries corresponding to the request.
         """
+        entries = self._get_entries_tuple(label_space_or_index)
+        if isinstance(entries, list):
+            return [entry.entry for entry in entries]
+        else:
+            return entries
```
Comment (akaszynski) [inline ansys/dpf/core/collection.py:206]: Unnecessary return.

```suggestion
        return entries
```

File: ansys/dpf/core/dpf_operator.py [hunk 7]
```diff
@@ -107,6 +115,17 @@ def _add_sub_res_operators(self, sub_results):
...
 
+    @property
+    @version_requires("3.0")
+    def progress_bar(self) -> bool:
+        """With this property, the user can choose to print a progress bar when
+        the operator's output is requested, default is False"""
+        return self._progress_bar
+
+    @progress_bar.setter
+    def progress_bar(self, value) -> bool:
```
Comment (akaszynski) [inline ansys/dpf/core/dpf_operator.py:126]: Setters shouldn't return a type hint other than `None`, but rather should have one for the argument.

```suggestion
    def progress_bar(self, value: bool) -> None:
```

See discussion:
https://github.com/python/mypy/issues/3004

File: ansys/dpf/core/dpf_operator.py [hunk 8]
```diff
@@ -144,14 +163,16 @@ def connect(self, pin, inpt, pin_out=0):
         request = operator_pb2.UpdateRequest()
         request.op.CopyFrom(self._message)
         request.pin = pin
-        _fillConnectionRequestMessage(request, inpt, pin_out)
+        tmp = _fillConnectionRequestMessage(request, inpt, self._server, pin_out)
```
Comment (akaszynski) [inline ansys/dpf/core/dpf_operator.py:166]: Not sure if `tmp` is necessary here.

File: ansys/dpf/core/dpf_operator.py [hunk 9]
```diff
@@ -144,14 +163,16 @@ def connect(self, pin, inpt, pin_out=0):
...
         if inpt is self:
             raise ValueError("Cannot connect to itself.")
         self._stub.Update(request)
 
     @protect_grpc
     def get_output(self, pin=0, output_type=None):
         """Retrieve the output of the operator on the pin number.
```
Comment (akaszynski) [inline ansys/dpf/core/dpf_operator.py:173]: Short and long description should have a line between.
```suggestion
        """Retrieve the output of the operator on the pin number.

```

File: ansys/dpf/core/dpf_operator.py [hunk 10]
```diff
@@ -144,14 +163,16 @@ def connect(self, pin, inpt, pin_out=0):
...
+        To activate the progress bar for server version higher or equal to 3.0,
+        use my_op.progress_bar = True
```
Comment (akaszynski) [inline ansys/dpf/core/dpf_operator.py:175]:
```suggestion
        use ``my_op.progress_bar=True``
```

File: ansys/dpf/core/dpf_operator.py [hunk 11]
```diff
@@ -478,6 +528,89 @@ def __truediv__(self, inpt):
...
+
+    def __str__(self):
+        out = ""
+        for key, i in self._asdict().items():
+            out += key + ": " + str(i) + "\n\n"
+        return out
+
+
+def available_operator_names(server=None):
+    """Returns the list of operators name available in the server
```
Comment (akaszynski) [inline ansys/dpf/core/dpf_operator.py:590]: Nitpick:
```suggestion
    """Returns the list of operator names available in the server.
```

File: ansys/dpf/core/dpf_operator.py [hunk 12]
```diff
@@ -478,6 +528,89 @@ def __truediv__(self, inpt):
...
+    -------
+    list
+
+    """
+    if server is None:
+        server = serverlib._global_server()
+    service = operator_pb2_grpc.OperatorServiceStub(server.channel).ListAllOperators(
+        operator_pb2.ListAllOperatorsRequest())
+    arr = []
+    import re
```
Comment (akaszynski) [inline ansys/dpf/core/dpf_operator.py:608]: Imports should be at the top of the module.

File: ansys/dpf/core/examples/downloads.py [hunk 13]
```diff
@@ -320,7 +316,7 @@ def download_fluent_files() -> dict:
 
 def download_extrapolation_3d_result() -> dict:
     """Download example static results of reference and integrated points
-    for extrapolation of 3d-element and return return the dictionary of 2 download paths.
+    for extrapolation of 3d-element and return return the dictionnary of 2 download paths.
```
Comment (akaszynski) [inline ansys/dpf/core/examples/downloads.py:319]:
```suggestion
    for extrapolation of 3d-element and return return the dictionary of 2 download paths.
```

File: ansys/dpf/core/examples/downloads.py [hunk 14]
```diff
@@ -333,7 +329,7 @@ def download_extrapolation_3d_result() -> dict:
 
     Examples
     --------
-    Download 2 example result files and return the dictionary containing 2 files
+    Download 2 example result files and return the dictionnary containing 2 files
```
Comment (akaszynski) [inline ansys/dpf/core/examples/downloads.py:332]:
```suggestion
    Download 2 example result files and return the dictionary containing 2 files
```

File: ansys/dpf/core/examples/downloads.py [hunk 15]
```diff
@@ -354,7 +350,7 @@ def download_extrapolation_3d_result() -> dict:
 
 def download_extrapolation_2d_result() -> dict:
     """Download example static results of reference and integrated points
-    for extrapolation of 2d-element and return the dictionary of 2 download paths.
+    for extrapolation of 2d-element and return the dictionnary of 2 download paths.
```
Comment (akaszynski) [inline ansys/dpf/core/examples/downloads.py:353]:
```suggestion
    for extrapolation of 2d-element and return the dictionary of 2 download paths.
```

File: ansys/dpf/core/examples/downloads.py [hunk 16]
```diff
@@ -367,7 +363,7 @@ def download_extrapolation_2d_result() -> dict:
 
      Examples
     --------
-    Download 2 example result files and return the dictionary containing 2 files
+    Download 2 example result files and return the dictionnary containing 2 files
```
Comment (akaszynski) [inline ansys/dpf/core/examples/downloads.py:366]:
```suggestion
    Download 2 example result files and return the dictionary containing 2 files
```

File: ansys/dpf/core/examples/downloads.py [hunk 17]
```diff
@@ -367,7 +363,7 @@ def download_extrapolation_2d_result() -> dict:
 
      Examples
```
Comment (akaszynski) [inline ansys/dpf/core/examples/downloads.py:364]:
```suggestion
    Examples
```

File: ansys/dpf/core/meshed_region.py [hunk 18]
```diff
@@ -470,6 +478,49 @@ def __send_init_request(self, num_nodes=0, num_elements=0):
...
+    def field_of_properties(self, property_name):
+        """Returns the ``Field`` or ``PropertyField`` associated
+        to a given property of the mesh
+
+        Parameters
+        ----------
+        property_name : str, common.elemental_properties, common.nodal_properties
+
+        Return
+        ------
```
Comment (akaszynski) [inline ansys/dpf/core/meshed_region.py:490]:
```suggestion
            Name of the property.

        Returns
        -------
```

File: ansys/dpf/core/meshes_container.py [hunk 19]
```diff
@@ -54,14 +54,11 @@ def plot(self, fields_container=None, **kwargs):
...
         >>> mesh = model.metadata.meshed_region
-        >>> split_mesh_op = dpf.Operator("split_mesh")
-        >>> split_mesh_op.connect(7, mesh)
-        >>> split_mesh_op.connect(13, "mat")
-        >>> meshes_cont = split_mesh_op.outputs.mesh_controller()
-        >>> disp_op = dpf.Operator("U")
-        >>> disp_op.connect(7, meshes_cont)
-        >>> ds = dpf.DataSources(examples.multishells_rst)
-        >>> disp_op.connect(4, ds)
+        >>> split_mesh_op = dpf.operators.mesh.split_mesh(mesh = mesh, property = "mat")
```
Comment (akaszynski) [inline ansys/dpf/core/meshes_container.py:57]:
```suggestion
        >>> split_mesh_op = dpf.operators.mesh.split_mesh(mesh=mesh, property="mat")
```

File: ansys/dpf/core/meshes_container.py [hunk 20]
```diff
@@ -54,14 +54,11 @@ def plot(self, fields_container=None, **kwargs):
...
+        >>> meshes_cont = split_mesh_op.eval()
+        >>> disp_op = dpf.operators.result.displacement(
+        ...     data_sources = dpf.DataSources(examples.multishells_rst),
+        ...     mesh = meshes_cont)
```
Comment (akaszynski) [inline ansys/dpf/core/meshes_container.py:61]: If wrapping lines, follow the [black code style](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html):
```suggestion
        ...     mesh = meshes_cont
        ... )
```

Also, don't use whitespace around `=` for kwargs.

File: ansys/dpf/core/operators/__init__.py [hunk 21]
```diff
@@ -1,22 +1,14 @@
-"""
-.. _ref_operators_package:
-
-ansys.dpf.core.operators
-========================
-"""
-  
-
 from . import result
```
Comment (akaszynski) [inline ansys/dpf/core/operators/__init__.py:1]: These all should be within parenthesis.

File: ansys/dpf/core/operators/averaging/elemental_difference.py [hunk 22]
```diff
@@ -1,248 +1,313 @@
...
+    >>> my_field = dpf.Field()
+    >>> op.inputs.field.connect(my_field)
+    >>> my_mesh_scoping = dpf.Scoping()
+    >>> op.inputs.mesh_scoping.connect(my_mesh_scoping)
+    >>> my_mesh = dpf.MeshedRegion()
+    >>> op.inputs.mesh.connect(my_mesh)
+    >>> my_through_layers = bool()
+    >>> op.inputs.through_layers.connect(my_through_layers)
+
+    >>> # Instantiate operator and connect inputs in one line
```
Comment (akaszynski) [inline ansys/dpf/core/operators/averaging/elemental_difference.py:50]: I know these are all autogenerated, but in the future comments should be on their own line without `>>> #` with whitespace between:

```suggestion
    Instantiate operator and connect inputs in one line

```

File: ansys/dpf/core/dpf_operator.py (hunk 8 above)
Comment (cbellot000) [reply ansys/dpf/core/dpf_operator.py:166]: it's necessary (but very ugly). I did that to make sure that tmp is not collected by the garbage collector

**Review by rlagha — APPROVED** [review]

**Pull request merged**